### PR TITLE
AMD64/Aarch64 compatible Dockerfile

### PR DESCRIPTION
**Pull request by joeabbey**
Using the Dockerfile setup from heighliner to enable amd64 and aarch64/arm64 builds when cgo is present.

 - [x] Remove the `installing rly binary... $(GOBIN)/rly` once this build passes

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,35 +1,83 @@
...
+        LIBDIR=/usr/x86_64-linux-musl/lib; \
+        mkdir -p $LIBDIR; \
+        export CC=x86_64-linux-musl-gcc CXX=x86_64-linux-musl-g++; \
+      fi; \
+    fi;
 
-ARG TARGETARCH TARGETOS
-ENV GOOS=${TARGETOS} GOARCH=${TARGETARCH}
-RUN make install
+RUN GOOS=linux GOARCH=$TARGETARCH CGO_ENABLED=1 LDFLAGS='-linkmode external -extldflags "-static"' make install
```
Comment (agouin) [inline Dockerfile]: Are CC and CXX vars needed for the make install? if so, this should be grouped with L16-L32. Looks like we don't need ARCH or the mkdir -p $LIBDIR. If we don't need CC and CXX vars for the make install, then we can remove L16-L32

Comment (joeabbey) [reply Dockerfile]: We almost certainly need the setting of the C/C++ compiler...

File: Dockerfile [hunk 2]
```diff
@@ -1,35 +1,82 @@
...
+        wget -c https://musl.cc/x86_64-linux-musl-cross.tgz -O - | tar -xzvv --strip-components 1 -C /usr; \
+    fi
 
-RUN go mod download
+ADD . .
 
-# Copy the files from host
-COPY . .
+RUN set -eux; \
+    LIBDIR=/lib; \
```
Comment (agouin) [inline Dockerfile]: we can remove LIBDIR, ARCH, and simplify the if/else to how it is L8-L12

File: Dockerfile [hunk 3]
```diff
@@ -1,35 +1,66 @@
...
-COPY go.mod go.sum ./
+ADD . .
 
-RUN go mod download
+RUN if [ "${TARGETARCH}" = "arm64" ] && [ "${BUILDARCH}" != "arm64" ]; then \
+        wget -c https://musl.cc/aarch64-linux-musl-cross.tgz -O - | tar -xzvv --strip-components 1 -C /usr; \
+        export CC=aarch64-linux-musl-gcc CXX=aarch64-linux-musl-g++;\
+    elif [ "${TARGETARCH}" = "amd64" ] && [ "${BUILDARCH}" != "amd64" ]; then \
+        wget -c https://musl.cc/x86_64-linux-musl-cross.tgz -O - | tar -xzvv --strip-components 1 -C /usr; \
+        export CC=x86_64-linux-musl-gcc CXX=x86_64-linux-musl-g++; \
```
Comment (agouin) [inline Dockerfile:19]: Let's keep the downloading of the toolchain step as a separate step before the `ADD . .`, this means we can use the cached toolchain (don't need to download the toolchain) every time there is a code change.

File: Dockerfile [hunk 4]
```diff
@@ -1,35 +1,72 @@
...
+    fi; \
+    GOOS=linux GOARCH=$TARGETARCH CGO_ENABLED=1 LDFLAGS='-linkmode external -extldflags "-static"' make install
 
-ARG TARGETARCH TARGETOS
-ENV GOOS=${TARGETOS} GOARCH=${TARGETARCH}
-RUN make install
+RUN if [ -d "/go/bin/linux_${TARGETARCH}" ]; then mv /go/bin/linux_${TARGETARCH}/* /go/bin/; fi
 
-FROM alpine:latest
+RUN if [ -d "/go/bin/linux_${TARGETARCH}" ]; then mv /go/bin/linux_${TARGETARCH}/* /go/bin/; fi
```
Comment (agouin) [inline Dockerfile]: duplicate of 23

File: Dockerfile (hunk 3 above)
Comment (joeabbey) [reply Dockerfile:19]: though... I wonder if we should just always install both toolchains.. 🤔

not sure how the layer caching is gonna handle two results from the same RUN line

Comment (agouin) [reply Dockerfile:19]: it will cache with it like this because of TARGETARCH

**Review by agouin — APPROVED** [review]
:partying_face: slim rly images!

**Pull request merged**